### DISCUSSION TITLE
TW be like Graph Databases

### DISCUSSION THREAD

**Discussion by daStretchyCat** [TiddlyWiki/TiddlyWiki5#7282] [Ideas]
TiddlyWiki is convenient for building a graph topology, and that's why I like it. 

https://github.com/rust-unofficial/awesome-rust/blob/main/README.md#database

Edit: I thought about comparing TW to some Graph DB. Existing Graph DB has access control, fancy visualization tools, scaling solutions (e.g., clustering), and so on. They are nice to have, but I'm not sure if it's viable to implement them.

**Comment by daStretchyCat** [comment]
[found the fancy visualization tool!](https://talk.tiddlywiki.org/t/echarts-bringing-amazing-visualization-tools-to-tiddlywiki/1944)

**Comment by Jermolene** [comment]
Hi @daStretchyCat it is fair to view TiddlyWiki as special kind of graph database. We even use indexes to optimise queries just like a real database.

The core was designed to allow other database engines to be dropped in, but there hasn't been a great deal of work on it. I'm particularly interested in using the WASM version of SQLite. The only requirement is that the core database must have a synchronous API. Databases that have an asynchronous API must instead be integrated via a syncer.

**Reply by rougsig** [reply]
> Databases that have an asynchronous API must instead be integrated via a syncer.

Could you possibly share an example or explanation of how to implement a syncer for integrating a database with an asynchronous API? Any code snippets, patterns, or libraries? TY

**Reply by MagicJF** [reply]
Somehow realted, probably you do find value on it:

https://github.com/TiddlyWiki/TiddlyWiki5/issues/8885#issuecomment-2585830436

**Reply by Arlen22** [reply]
@rougsig just saw this, and thought I'd mention that it's called a sync adaptor. Take a look at `plugins/tiddlywiki/tiddlyweb` and `plugins/tiddlywiki/filesystem` for examples. There are also several solutions already available in the community, but I'm not familiar enough with them yet.

**Reply by rougsig** [reply]
@Arlen22 Thank you. For now i started pet-project tol solve my goals. I have plans to use TW5, but without ability to merge multiple datasources into one, it can't solve my goals.
